source/opt/combine_access_chains.cpp

```diff
@@ -0,0 +1,253 @@
+    Instruction* new_value_inst =
+        constant_mgr->GetDefiningInstruction(new_value_constant);
+    new_value_id = new_value_inst->result_id();
+  } else if (!type->AsStruct() || combining_element_operands) {
+    // TODO(alan-baker): handle this unlikely case.
+    if (last_index_inst->type_id() != element_inst->type_id()) return false;
+
+    // Generate an addition of the two indices.
+    InstructionBuilder builder(
+        context(), inst,
+        IRContext::kAnalysisDefUse | IRContext::kAnalysisInstrToBlockMapping);
+    Instruction* addition = builder.AddIAdd(last_index_inst->type_id(),
+                                            last_index_inst->result_id(),
+                                            element_inst->result_id());
+    new_value_id = addition->result_id();
+  } else {
+    // Indexing into structs must be constant, so bail out here.
+    return false;
+  }
+  new_operands->push_back({SPV_OPERAND_TYPE_ID, {new_value_id}});
+  return true;
+}
+
+bool CombineAccessChains::CreateNewOperands(
+    Instruction* ptr_input, Instruction* inst,
+    std::vector<Operand>* new_operands) {
+  // Start by copying all the input operands of the feeder access chain.
+  for (uint32_t i = 0; i != ptr_input->NumInOperands() - 1; ++i) {
+    new_operands->push_back(ptr_input->GetInOperand(i));
+  }
+
+  // Deal with the last index of the feeder access chain.
+  if (inst->opcode() == SpvOpPtrAccessChain) {
+    // The last index of the feeder should be combined with the element operand
+    // of |inst|.
+    if (!CombineIndices(ptr_input, inst, new_operands)) return false;
+  } else {
+    // The indices aren't being combined so now add the last index operand of
+    // |ptr_input|.
+    new_operands->push_back(
+        ptr_input->GetInOperand(ptr_input->NumInOperands() - 1));
+  }
+
+  // Copy the remaining index operands.
+  uint32_t starting_index = (inst->opcode() == SpvOpPtrAccessChain) ? 2 : 1;
+  for (uint32_t i = starting_index; i < inst->NumInOperands(); ++i) {
+    new_operands->push_back(inst->GetInOperand(i));
+  }
+
+  return true;
+}
+
+bool CombineAccessChains::CombineAccessChain(Instruction* inst) {
+  assert((inst->opcode() == SpvOpPtrAccessChain ||
+          inst->opcode() == SpvOpAccessChain ||
+          inst->opcode() == SpvOpInBoundsAccessChain) &&
+         "Wrong opcode. Expected an access chain.");
+
+  Instruction* ptr_input =
+      context()->get_def_use_mgr()->GetDef(inst->GetSingleWordInOperand(0));
+  if (ptr_input->opcode() != SpvOpAccessChain &&
+      ptr_input->opcode() != SpvOpInBoundsAccessChain &&
+      ptr_input->opcode() != SpvOpPtrAccessChain) {
+    return false;
+  }
+
+  // Handles the following cases:
+  // 1. |inst| is not a pointer access chain.
+  //    |inst|'s indices are appended to |ptr_input|'s indices.
+  // 2. |ptr_input| is not pointer access chain.
+  //    |inst|'s element operand is combined with the last index in
```

Comment:
In 2. you should be more clear to state that |inst| *is* a ptr access chain.  (It's not clear that 2 is an "else" clause of 1.)  
It's kind of implied by calling out inst's "element" operand (which exists in PtrAccessChain but not AccessChain), but that's easy to miss.